libvast/src/system/evaluator.cpp

```diff
@@ -0,0 +1,173 @@
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/system/evaluator.hpp"
+
+#include <caf/actor.hpp>
+#include <caf/behavior.hpp>
+#include <caf/event_based_actor.hpp>
+#include <caf/stateful_actor.hpp>
+
+#include "vast/expression_visitors.hpp"
+#include "vast/logger.hpp"
+#include "vast/system/atoms.hpp"
+
+namespace vast::system {
+
+namespace {
+
+/// Concatenates IDs according to given predicates. In paticular, resolves
+/// conjunctions, disjunctions, and negations.
+struct ids_evaluator {
+  ids operator()(caf::none_t) const {
+    return {};
+  }
+
+  ids operator()(const conjunction& c) const {
+    auto result = caf::visit(*this, c[0]);
+    if (result.empty() || all<0>(result))
+      return {};
+    for (size_t i = 1; i < c.size(); ++i) {
+      result &= caf::visit(*this, c[i]);
+      // This short-circuit works if and only if all bitmaps have the same size.
+      // if (result.empty() || all<0>(result)) // short-circuit
+      //   return {};
+    }
+    return result;
+  }
+
+  ids operator()(const disjunction& d) const {
+    ids result;
+    for (auto& op : d) {
+      result |= caf::visit(*this, op);
+      // This short-circuit works if and only if all bitmaps have the same size.
+      // if (all<1>(result)) // short-circuit
+      //   break;
+    }
+    return result;
+  }
+
+  ids operator()(const negation& n) const {
+    auto result = caf::visit(*this, n.expr());
+    result.flip();
+    return result;
+  }
+
+  ids operator()(const predicate& pred) const {
+    auto i = xs.find(pred);
+    return i != xs.end() ? i->second.second : ids{};
+  }
+
+  const evaluator_state::sub_hits_map& xs;
+};
+
+} // namespace
+
+evaluator_state::evaluator_state(caf::event_based_actor* self) : self(self) {
+  // nop
+}
+
+void evaluator_state::init(caf::actor client, expression expr) {
+  this->client = std::move(client);
+  this->expr = std::move(expr);
+}
+
+void evaluator_state::handle_indexer_result(const predicate& pred,
+                                            const ids& result) {
+  VAST_DEBUG(self, "got new hits", result, " for predicate", pred);
+  if (auto ptr = sub_hits_of(pred); ptr == nullptr) {
+    VAST_ERROR(self, "got a result for an unknown predicate:", pred);
+  } else {
+    auto& [missing, accumulated_hits] = *ptr;
+    accumulated_hits |= result;
+    if (--missing == 0) {
+      VAST_DEBUG(self, "collected all INDEXER results for predicate", pred);
+      update();
+    }
+  }
+  decrement_pending();
+}
+
+void evaluator_state::handle_missing_indexer_result(const predicate& pred,
+                                                    const caf::error& err) {
+  VAST_IGNORE_UNUSED(err);
+  VAST_DEBUG(self, "got no result for predicate", pred, "INDEXER returned",
+             self->system().render(err));
```

Comment:
Some punctuation in this debug statement is needed to separate the two sentences.